File: 3.x.x/crates-utopia/queue/src/broker/pool.rs
use std::sync::Arc;

use serde_json::Value;

use crate::consumer::Consumer;
use crate::error::QueueError;
use crate::message::Message;
use crate::pool::ResourcePool;
use crate::publisher::Publisher;
use crate::queue::Queue;

/// Checkout a publisher/consumer from in-memory pools.
///
/// PHP `Utopia\Queue\Broker\Pool`.
pub struct Pool {
    publisher: Option<Arc<ResourcePool<Arc<dyn Publisher>>>>,
    consumer: Option<Arc<ResourcePool<Arc<dyn Consumer>>>>,
}

impl std::fmt::Debug for Pool {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("Pool")
            .field("publisher", &self.publisher.is_some())
            .field("consumer", &self.consumer.is_some())
            .finish()
    }
}

impl Pool {
    pub fn new(
        publisher: Option<Arc<ResourcePool<Arc<dyn Publisher>>>>,
        consumer: Option<Arc<ResourcePool<Arc<dyn Consumer>>>>,
    ) -> Self {
        Self {
            publisher,
            consumer,
        }
    }

    pub fn from_publisher_pool(pool: ResourcePool<Arc<dyn Publisher>>) -> Self {
        let pool = Arc::new(pool);
        Self {
            publisher: Some(pool.clone()),
            consumer: None,
        }
    }
}

impl Publisher for Pool {
    fn enqueue(&self, queue: &Queue, payload: Value, priority: bool) -> Result<bool, QueueError> {
        self.publisher
            .as_ref()
            .map(|p| p.use_item(|b| b.enqueue(queue, payload, priority)))
            .transpose()?
            .ok_or_else(|| QueueError::Other("publisher pool is not configured".into()))
    }

    fn retry(
        &self,
        queue: &Queue,
        limit: Option<i64>,
        max_attempts: Option<i64>,
        newer_than: Option<i64>,
    ) -> Result<(), QueueError> {
        if let Some(p) = &self.publisher {
            p.use_item(|b| b.retry(queue, limit, max_attempts, newer_than))?;
        }
        Ok(())
    }

    fn get_queue_size(&self, queue: &Queue, failed_jobs: bool) -> Result<i64, QueueError> {
        self.publisher
            .as_ref()
            .map(|p| p.use_item(|b| b.get_queue_size(queue, failed_jobs)))
            .transpose()?
            .ok_or_else(|| QueueError::Other("publisher pool is not configured".into()))
    }

    fn reap(
        &self,
        queue: &Queue,
        older_than: i64,
        limit: Option<i64>,
        max_attempts: Option<i64>,
        newer_than: Option<i64>,
    ) -> Result<i64, QueueError> {
        self.publisher
            .as_ref()
            .map(|p| p.use_item(|b| b.reap(queue, older_than, limit, max_attempts, newer_than)))
            .transpose()?
            .ok_or_else(|| QueueError::Other("publisher pool is not configured".into()))
    }
}

impl Consumer for Pool {
    fn receive(&self, queue: &Queue, timeout: i64) -> Result<Option<Message>, QueueError> {
        self.consumer
            .as_ref()
            .map(|p| p.use_item(|b| b.receive(queue, timeout)))
            .transpose()?
            .ok_or_else(|| QueueError::Other("consumer pool is not configured".into()))
    }

    fn commit(&self, queue: &Queue, message: &Message) -> Result<(), QueueError> {
        if let Some(p) = &self.consumer {
            p.use_item(|b| b.commit(queue, message))?;
        }
        Ok(())
    }

    fn reject(&self, queue: &Queue, message: &Message) -> Result<(), QueueError> {
        if let Some(p) = &self.consumer {
            p.use_item(|b| b.reject(queue, message))?;
        }
        Ok(())
    }

    fn close(&self) {
        // PHP: TODO close all connections in the pool.
    }

    fn as_publisher(&self) -> Option<&dyn Publisher> {
        if self.publisher.is_some() {
            Some(self)
        } else {
            None
        }
    }
}

File: src/Queue/Broker/Pool.php
<?php

namespace Utopia\Queue\Broker;

use Utopia\Pools\Pool as UtopiaPool;
use Utopia\Queue\Consumer;
use Utopia\Queue\Message;
use Utopia\Queue\Publisher;
use Utopia\Queue\Queue;

readonly class Pool implements Publisher, Consumer
{
    public function __construct(
        private ?UtopiaPool $publisher = null,
        private ?UtopiaPool $consumer = null,
    ) {}

    public function enqueue(Queue $queue, array $payload, bool $priority = false): bool
    {
        return $this->delegate($this->publisher, __FUNCTION__, \func_get_args());
    }

    public function enqueueMany(Queue $queue, array $payloads, bool $priority = false): bool
    {
        return $this->delegate($this->publisher, __FUNCTION__, \func_get_args());
    }

    public function retry(Queue $queue, ?int $limit = null, ?int $maxAttempts = null, ?int $newerThan = null): void
    {
        $this->delegate($this->publisher, __FUNCTION__, \func_get_args());
    }

    /**
     * {@see Redis::reap()} — requires the pooled publisher to be a broker that
     * implements it.
     */
    public function reap(Queue $queue, int $olderThan = 90000, ?int $limit = null, ?int $maxAttempts = null, ?int $newerThan = null): int
    {
        return $this->delegate($this->publisher, __FUNCTION__, \func_get_args());
    }

    public function getQueueSize(Queue $queue, bool $failedJobs = false): int
    {
        return $this->delegate($this->publisher, __FUNCTION__, \func_get_args());
    }

    public function receive(Queue $queue, int $timeout): ?Message
    {
        return $this->delegate($this->consumer, __FUNCTION__, \func_get_args());
    }

    public function commit(Queue $queue, Message $message): void
    {
        $this->delegate($this->consumer, __FUNCTION__, \func_get_args());
    }

    public function reject(Queue $queue, Message $message): void
    {
        $this->delegate($this->consumer, __FUNCTION__, \func_get_args());
    }

    /**
     * Tell the broker the handler is still working on this message.
     *
     * Present so a pooled consumer does not silently lose ack extension.
     * {@see \Utopia\Queue\Adapter\Swoole::withAckExtension()} probes the
     * consumer it was handed, which for a pooled worker is this class and not
     * the broker underneath -- so without these two methods the heartbeat found
     * nothing to call and quietly degraded to the behaviour it exists to fix:
     * ackWait back to being a hard ceiling on job duration, with a redelivery
     * running concurrently with the attempt still in progress.
     *
     * Capability is probed per lease rather than declared, because the pool may
     * hold a broker with no notion of extension (Redis), and this class is the
     * consumer for those too.
     *
     * Correct while the pool leases the same broker to a message's receive()
     * and its extend() -- true at size 1, which is the documented wiring, and
     * the same assumption commit() and reject() already make. At a larger size
     * a lease can land on a broker that never held this message, whose in-flight
     * map has no entry for it; {@see \Utopia\Queue\Broker\Nats::extend()} is
     * silent in that case by design.
     */
    public function extend(Queue $queue, Message $message): void
    {
        $this->consumer?->use(function (Publisher|Consumer $adapter) use ($queue, $message): void {
            $extend = [$adapter, 'extend'];

            if (\is_callable($extend)) {
                $extend($queue, $message);
            }
        });
    }

    /**
     * How often {@see self::extend()} should be called, or null when the leased
     * broker cannot extend at all.
     *
     * Null rather than a fallback number: there is no interval that is safe to
     * guess. One longer than the real ackWait extends nothing while looking like
     * it does, which is the same silent degradation in a more convincing
     * costume. The caller skips the heartbeat entirely instead.
     */
    public function extendInterval(): ?float
    {
        return $this->consumer?->use(function (Publisher|Consumer $adapter): ?float {
            $interval = [$adapter, 'extendInterval'];

            return \is_callable($interval) ? (float) $interval() : null;
        });
    }

    /**
     * Run idle upkeep across both pools.
     *
     * Safe to call at any time, including while a receive loop is running: the
     * sweep reaches only resources sitting idle in the pool, never the one a
     * caller currently holds. That is the whole difference from calling
     * {@see Nats::tick()} directly, which needs exclusive access.
     *
     * A pooled broker is the case the keepalive exists for — a publisher slot
     * can sit untouched for longer than the server's ping deadline and be
     * reaped without anyone noticing until the next publish.
     */
    public function maintain(): void
    {
        $this->publisher?->maintain();

        // Distinct pools; a broker configured with only one leaves the other null.
        if ($this->consumer instanceof UtopiaPool && $this->consumer !== $this->publisher) {
            $this->consumer->maintain();
        }
    }

    public function close(): void
    {
        // TODO: Implement closing all connections in the pool
    }

    /**
     * @param array<mixed> $args
     */
    protected function delegate(?UtopiaPool $pool, string $method, array $args): mixed
    {
        return $pool?->use(fn(Publisher|Consumer $adapter) => $adapter->$method(...$args));
    }
}
